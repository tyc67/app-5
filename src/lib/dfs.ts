import { type Node } from './nodeProcessor'
import { getNeighbors } from './neighborsNode'

// const directions: [dx: number, dy: number][] = [[1,0],[0,1], [-1,0], [0,-1]]

// async function dfs(node: Node, matrix: Node[][], endNode: Node, path: Node[], onVisit: (row: number, col: number) => void) {
//   let answer: Node[] | null = null
//   if(node === endNode && path.length) {
//     answer = [...path]
//   }

//   node.isVisited = true
//   onVisit(node.row, node.col)
//   path.push(node);

//   directions.forEach(([dx, dy]) => {
//     const neighbor = matrix[node.row + dx][node.col + dy]
//     if (neighbor.isBlock || neighbor.isVisited) return
//     if (!answer) {
//       answer = dfs(neighbor, matrix, endNode, path)
//     }
//   })

//   path.pop();

//   return answer

// }

export default function DFS(nodes: Node[][], startNode: Node, endNode: Node) {
  const stack: Node[] = []
  const visitedNodes = []
  const shortestPath: Node[] = []
  stack.push(startNode)
  while (stack.length > 0) {
    const currentNode = stack.pop()
    if (currentNode) {
      currentNode.isVisited = true
      visitedNodes.push(currentNode)
      if (currentNode === endNode) {
        return { visitedNodes, shortestPath }
      }

      const neighbors = getNeighbors(currentNode, nodes)
      for (const neighbor of neighbors) {
        if (!neighbor.isVisited) {
          stack.push(neighbor)
        } else {
          neighbor.previousNode = currentNode
        }
      }
    }
  }

  return { visitedNodes, shortestPath }
}
