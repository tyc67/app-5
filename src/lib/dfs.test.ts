import dfs from './dfs'

describe('dfs', () => {
  it('should find a path as expected', () => {
    const matrix = [
      [
        {
          col: 0,
          row: 0,
          distance: 1,
          opaque: 0,
          isEnd: false,
          isStart: true,
          isVisited: false,
          isVisualized: false,
          isPath: false,
          previousNode: null,
        },
        {
          col: 0,
          row: 1,
          distance: 1,
          opaque: 0,
          isEnd: true,
          isStart: false,
          isVisited: false,
          isVisualized: false,
          isPath: false,
          previousNode: null,
        },
      ],
    ]
    const actual = dfs(matrix, matrix[0][0], matrix[0][1])
    expect(actual.shortestPath).toEqual([matrix[0][0], matrix[0][1]])
    expect(actual.visitedNodes).toEqual([matrix[0][0], matrix[0][1]])
  })
})
