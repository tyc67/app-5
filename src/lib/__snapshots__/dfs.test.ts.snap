// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dfs should find a path as expected 1`] = `[]`;
